File: python/mlp_xor.py
import numpy as np
from util import *

# XOR
X = np.array([
  [0,1],
  [1,0],
  [0,0],
  [1,1]
])

T = np.array([
  [1],
  [1],
  [0],
  [0]
])

np.random.seed(34)

# 重み初期化
# lecun_uniform
# http://yann.lecun.com/exdb/publis/pdf/lecun-98b.pdf
w1 = np.random.uniform(-np.sqrt(1.0/2), np.sqrt(1.0/2), (2, 8)).astype('float32')
b1 = np.zeros(8).astype('float32')
w2 = np.random.uniform(-np.sqrt(1.0/8), np.sqrt(1.0/8), (8, 1)).astype('float32')
b2 = np.zeros(1).astype('float32')

def forward(x, w1, b1, w2, b2):
    u1 = np.matmul(x, w1) + b1
    h1 = relu(u1)
    u2 = np.matmul(h1, w2) + b2
    y = sigmoid(u2)
    return u1, h1, u2, y


# train
for epoch in range(0, 3000):
    for x, t in zip(X, T):
        # forward
        u1, h1, u2, y = forward(x, w1, b1, w2, b2)

        # backprop
        delta2 = y - t
        delta1 = np.matmul(delta2, w2.T) * deriv_relu(u1)

        # 勾配
        dW2 = np.matmul(h1[None, :].T, delta2[None, :])
        dW1 = np.matmul(x[None, :].T, delta1[None, :])
        db2 = delta2
        db1 = delta1

        # 重み更新　
        lr = 0.05
        w2 -= lr * dW2
        w1 -= lr * dW1
        b2 -= lr * db2
        b1 -= lr * db1


# pred
for x in X:
    _, _, _, y = forward(x, w1, b1, w2, b2)
    print(x,y)


#
# gradient checking
#

# np.logでdivide by zero encountered in logエラーが出るのを回避するためのヘルパ
# aの最小値を決めて分母が0になるのを避ける
def np_log(a):
    return np.log(np.clip(a, 1e-10, a))

# 2値分類なので、尤度関数 = y(t) * (1-y)(1-t) yは確率。t={0,1}
# この尤度関数を最大化したい => 積は面倒なのでlogをとって和の関数にする。符号をマイナスに
# cost = log(y(t) * (1-y)(1-t))
# cost = log(y(t)) + log((1-y)(1-t))
# cost = {t*log(y) + (1-t)*log(1-y)}
# cost = t*log(y) + (1-t)*log(1-y)}
# cost = - { t*log(y) + (1-t)*log(1-y)} }
def calc_cost(t, y):
    return - (t * np_log(y) + (1 - t) * np_log(1 - y))

w1 = np.random.uniform(-0.08, 0.08, (2, 8)).astype('float32')
b1 = np.zeros(8).astype('float32')
w2 = np.random.uniform(-0.08, 0.08, (8, 1)).astype('float32')
b2 = np.zeros(1).astype('float32')
for x, t in zip(X, T):
    #
    # compute graditents by back propagation
    #

    # forward
    u1, h1, u2, y = forward(x, w1, b1, w2, b2)

    # backprop
    delta2 = y - t
    delta1 = np.matmul(delta2, w2.T) * deriv_relu(u1)

    # 勾配
    dW2 = np.matmul(h1[None, :].T, delta2[None, :])
    dW1 = np.matmul(x[None, :].T, delta1[None, :])
    db2 = delta2
    db1 = delta1

    #
    # compute graditents by numerical grad
    #

    # forward
    _, _, _, y = forward(x, w1, b1, w2, b2)

    # calc numerical grad
    # コスト関数をw1,b1,w2,b2でそれぞれ数値微分（+h,-hでそれぞれ計算する）して勾配を求める
    h = 1e-5 # 0.00001

    # w1
    # w1（行列）の値を一個ずつ+h,-hで微分していく
    n_dW1 = np.zeros_like(w1)
    for i, w1_row in enumerate(w1):
        for j, _ in enumerate(w1_row):
            # i,jの位置の値だけ+hした重みで予測をしてcost_hを取得し、上述のcostとの差をとって微分
            tmp = w1[i][j]
            # plus
            w1[i][j] = tmp + h
            _, _, _, y = forward(x, w1, b1, w2, b2)
            cost_plus = calc_cost(t, y)
            # minus
            w1[i][j] = tmp - h
            _, _, _, y = forward(x, w1, b1, w2, b2)
            cost_minus = calc_cost(t, y)
            # grad
            n_dW1[i][j] = (cost_plus - cost_minus) / (2 * h)
            w1[i][j] = tmp  # 元に戻す

    # db1
    n_db1 = np.zeros_like(b1)
    for i, _ in enumerate(b1):
        tmp = b1[i]
        # plus
        b1[i] = tmp + h
        _, _, _, y = forward(x, w1, b1, w2, b2)
        cost_plus = calc_cost(t, y)
        # minus
        b1[i] = tmp - h
        _, _, _, y = forward(x, w1, b1, w2, b2)
        cost_minus = calc_cost(t, y)
        # grad
        n_db1[i] = (cost_plus - cost_minus) / (2 * h)
        b1[i] = tmp  # 元に戻す

    # dw2
    n_dW2 = np.zeros_like(w2)
    for i, w2_row in enumerate(w2):
        for j, _ in enumerate(w2_row):
            tmp = w2[i][j]
            w2[i][j] = tmp + h
            _, _, _, y = forward(x, w1, b1, w2, b2)
            cost_plus = calc_cost(t, y)
            w2[i][j] = tmp - h
            _, _, _, y = forward(x, w1, b1, w2, b2)
            cost_minus = calc_cost(t, y)
            n_dW2[i][j] = (cost_plus - cost_minus) / (2 * h)
            w2[i][j] = tmp

    # db2
    n_db2 = np.zeros_like(b2)
    for i, _ in enumerate(b2):
        tmp = b2[i]
        b2[i] = tmp + h
        _, _, _, y = forward(x, w1, b1, w2, b2)
        cost_plus = calc_cost(t, y)
        b2[i] = tmp - h
        cost_minus = calc_cost(t, y)
        n_db2[i] = (cost_plus - cost_minus) / (2 * h)
        b2[i] = tmp

    #print("dW1:", dW1)
    #print("n_dW1:", n_dW1)
    #print("db1:", db1)
    #print("n_db1:", n_db1)
    #print("dW2:", dW2)
    #print("n_dW2:", n_dW2)
    #print("db2:", db2)
    #print("n_db2:", n_db2)
    diff = np.average(np.abs(dW1 - n_dW1))
    print("diff: %.10f" % diff)

File: ruby/mlp_xor.rb
#
# XOR classifier
#

require 'pp'
require './util'

# XOR
X = [
  [0,1],
  [1,0],
  [0,0],
  [1,1]
]
T = [
  [1],
  [1],
  [0],
  [0]
]

srand(34)
w1 = create_weight(2, 8)
b1 = create_bias(8)
w2 = create_weight(8, 1)
b2 = create_bias(1)

# 注意: trainのところと同じになるように
def forward(x, w1, w2, b1, b2)
  u1 = matmul(x, w1)
  u1 = u1.zip(b1).map { |u,b| u + b }
  h1 = u1.map {|v| relu(v) }
  u2 = matmul(h1, w2)
  u2 = u2.zip(b2).map { |u,b| u + b }
  y = u2.map {|v| sigmoid(v)}
  [u1, h1, u2, y]
end

(0...3000).each do |epoch|
  X.zip(T).each do |x, t|
    # forward
    u1, h1, u2, y = forward(x, w1, w2, b1, b2)

    # backprop
    # y,tは配列（要素数1）
    delta2 = y.zip(t).map { |yy, tt| yy - tt }
    a = matmul(delta2, w2.transpose)
    delta1 = a.zip(u1).map { |aa, uu1| aa * deriv_relu(uu1) }

    # 勾配
    dw1 = vv(x, delta1)
    dw2 = vv(h1, delta2)
    db1 = delta1
    db2 = delta2

    # 重み更新　
    lr = 0.05
    w1 = mat_minus(w1, dw1, lr)
    w2 = mat_minus(w2, dw2, lr)
    b1 = b1.zip(db1).map { |b, d| b - lr * d }
    b2 = b2.zip(db2).map { |b, d| b - lr * d }
  end
end



X.each do |x|
  _, _, _, y = forward(x, w1, w2, b1, b2)
  y = y[0]
  p "#{x}: " + sprintf("%.5f", y)
end

#
# gradient checking
#

def calc_cost(t, y)
  # maximize: y^t * (1-y)^(1-t)
  # =>
  # minimize: log(y^t * (1-y)^(1-t))
  # = log(y^t) + log(1-y)^(1-t)
  # = t * log(y) + (1-t) * log(1-y)
  - (t * Math.log(y) + (1-t) * Math.log(1-y))
end

# weights and biases for gradient che3cking
w1 = create_weight(2, 8)
b1 = create_bias(8)
w2 = create_weight(8, 1)
b2 = create_bias(1)

# data for gradient checking
x = X[0]
t = T[0]

#
# compute graditents by back propagation
#
u1, h1, u2, y = forward(x, w1, w2, b1, b2)
delta2 = y.zip(t).map {|yy,tt| yy - tt}
a = matmul(delta2, w2.transpose)
delta1 = a.zip(u1).map {|aa, uu1| aa * deriv_relu(uu1)}
dw2 = vv(h1, delta2)
dw1 = vv(x, delta1)
db2 = delta2
db1 = delta1

#
# compute graditents by numerical grad
#

h = 1e-5

# w2
n_dw2 = w2.map.with_index do |row, i|
  row.map.with_index do |v, j|
    w2[i][j] = v + h
    _, _, _, y = forward(x, w1, w2, b1, b2)
    cost_plus = calc_cost(t[0], y[0])
    w2[i][j] = v - h
    _, _, _, y = forward(x, w1, w2, b1, b2)
    cost_minus = calc_cost(t[0], y[0])
    grad = (cost_plus - cost_minus) / (2 * h)
    w2[i][j] = v # 元に戻す
    grad
  end
end

# w1
n_dw1 = w1.map.with_index do |row, i|
  row.map.with_index do |v, j|
    w1[i][j] = v + h
    _, _, _, y = forward(x, w1, w2, b1, b2)
    cost_plus = calc_cost(t[0], y[0])
    w1[i][j] = v - h
    _, _, _, y = forward(x, w1, w2, b1, b2)
    cost_minus = calc_cost(t[0], y[0])
    grad = (cost_plus - cost_minus) / (2 * h)
    w1[i][j] = v
    grad
  end
end

# b2
# b1
#
p "dw2: #{dw2}"
p "n_dw2: #{n_dw2}"

p "dw1: #{dw1}"
p "n_dw1: #{n_dw1}"
